protocols/autonat/src/behaviour.rs

```diff
@@ -298,43 +302,44 @@ impl Behaviour {
             ongoing_inbound: &mut self.ongoing_inbound,
         }
     }
-}
-
-impl NetworkBehaviour for Behaviour {
-    type ConnectionHandler = <RequestResponse<AutoNatCodec> as NetworkBehaviour>::ConnectionHandler;
-    type OutEvent = Event;
 
-    fn inject_connection_established(
+    /// Called on the [`FromSwarm::ConnectionEstablished`]
+    /// [event](`NetworkBehaviour::on_swarm_event`).
+    fn on_connection_established(
         &mut self,
-        peer: &PeerId,
-        conn: &ConnectionId,
-        endpoint: &ConnectedPoint,
-        failed_addresses: Option<&Vec<Multiaddr>>,
-        other_established: usize,
-    ) {
-        self.inner.inject_connection_established(
-            peer,
-            conn,
+        ConnectionEstablished {
+            peer_id,
+            connection_id,
             endpoint,
             failed_addresses,
             other_established,
-        );
-        let connections = self.connected.entry(*peer).or_default();
+        }: ConnectionEstablished,
+    ) {
+        self.inner
+            .on_swarm_event(FromSwarm::ConnectionEstablished(ConnectionEstablished {
+                peer_id,
+                connection_id,
+                endpoint,
+                failed_addresses,
+                other_established,
+            }));
```

Comment:
> we can for the variants that can be `Copy` which is all but `ConnectionClosed`. Did it

Hmm, it would be really nice if the entire event is `Clone` but as you said, `ConnetionClosed` can't be because of `Handler`. That is quite unfortunate. I wonder if we can work around this somehow. Having to unpack every `FromSwarm` because you want to delegate to an inner implementation is really annoying. The fact that this composition is not completely smooth may bite us in the future :/

I'll track this separately, no need to block this PR here :)